packages/jsii/lib/transforms/deprecated-warnings-transformer.ts

```diff
@@ -0,0 +1,258 @@
+              ts.createAssignment(
+                ts.createIdentifier('const message'),
+                ts.createLiteral(
+                  `${warning?.elementName} is deprecated.\n  ${warning?.message}\n  This API will be removed in the net major release.`,
+                ),
+              ),
+            ),
+            // ts.createExpressionStatement(
+            ts.createSwitch(
+              ts.createIdentifier('deprecationMode'),
+              ts.createCaseBlock([
+                ts.createCaseClause(ts.createLiteral('error'), [
+                  ts.createThrow(
+                    ts.createNew(ts.createIdentifier('Error'), [], []),
+                  ),
+                  ts.createBreak(),
+                ]),
+                ts.createCaseClause(ts.createLiteral('warn'), [
+                  ts.createExpressionStatement(
+                    ts.createCall(
+                      ts.createIdentifier('console.warn'),
+                      [],
+                      [ts.createLiteral(message)],
+                    ),
+                  ),
+                  ts.createBreak(),
+                ]),
+              ]),
+            ),
+          ],
+          true,
+        ),
+      ),
+      ts.createExpressionStatement(
+        ts.createCall(ts.createIdentifier('printWarning'), [], []),
+      ),
+    ];
+  }
+
+  private findDeprecated(toProcess: ts.Type[]): ts.Type | undefined {
+    if (toProcess.length === 0) {
+      return undefined;
+    }
+
+    if (this.deprecatedTypes.includes(toProcess[0])) {
+      return toProcess[0];
+    }
+
+    const node = this.index.get(
+      this.typeChecker.getFullyQualifiedName(toProcess[0].symbol),
+    )!;
+    if (!ts.isClassDeclaration(node) && !ts.isInterfaceDeclaration(node)) {
+      throw new Error('Node should be a class or interface declaration.');
+    }
+
+    const newBatch = toProcess.slice(1);
+    const heritageClauses = node.heritageClauses ?? [];
+    for (const clause of heritageClauses) {
+      for (const expression of clause.types) {
+        const type = this.typeChecker.getTypeAtLocation(expression);
+        if (!toProcess.includes(type)) {
+          newBatch.push(type);
+        }
+      }
+    }
+
+    return this.findDeprecated(newBatch);
+  }
+}
+
+function isMethodLikeDeclaration(node: ts.Node): boolean {
+  return (
+    ts.isMethodDeclaration(node) ||
+    ts.isGetAccessorDeclaration(node) ||
+    ts.isSetAccessorDeclaration(node) ||
+    ts.isConstructorDeclaration(node)
+  );
+}
+
+function getWarning(
+  typeChecker: ts.TypeChecker,
+  node: ts.Node,
+): Warning | undefined {
+  const original = ts.getOriginalNode(node);
+  const warnings = ts
+    .getJSDocTags(original)
+    .filter(
+      (tag) => (tag.tagName.text ?? tag.tagName.escapedText) === 'deprecated',
+    )
+    .map((tag: ts.JSDocTag) => {
+      const type = typeChecker.getTypeAtLocation(node);
+      // TODO We need a better fully qualified name
```

Comment:
The name that gets generated here is something like `"/Users/otaviom/projects/experiments/jsii-experiment/index".MyClass`. But we probably want the JSII style: `jsii-experiment.MyClass`.